File: specs/009-kafka-events/contracts/event_schemas.py
# Kafka Event Streaming - Pydantic Schemas

> **Language**: Python 3.13+
> **Purpose**: Event schema contracts for validation and serialization
> **Status**: Production Ready | **Deployment**: Not Required

```python
"""
Kafka Event Schema Contracts for Todo List Application

This module defines Pydantic models for all events in the system.
These schemas ensure type safety, validation, and serialization consistency.

Usage:
    from backend.services.events.schemas import TaskCreatedEvent, BaseEvent

    event = TaskCreatedEvent(
        event_id=uuid4(),
        timestamp=datetime.now(timezone.utc),
        correlation_id=uuid4(),
        producer="chat-api",
        payload=TaskCreatedPayload(...)
    )

    # Validate and serialize
    event_json = event.model_dump(mode="json")
"""

from datetime import datetime, timezone
from enum import Enum
from typing import Literal, Union
from uuid import UUID, uuid4

from pydantic import BaseModel, Field, field_validator


# ============================================================================
# ENUMS
# ============================================================================

class EventType(str, Enum):
    """All possible event types in the system."""
    TASK_CREATED = "task-created"
    TASK_UPDATED = "task-updated"
    TASK_DELETED = "task-deleted"
    TASK_COMPLETED = "task-completed"
    REMINDER_SCHEDULED = "reminder-scheduled"
    REMINDER_TRIGGERED = "reminder-triggered"
    REMINDER_CANCELLED = "reminder-cancelled"
    RECURRENCE_INSTANCE_CREATED = "recurrence-instance-created"


class EventProducer(str, Enum):
    """Services that can produce events."""
    CHAT_API = "chat-api"
    REMINDER_WORKER = "reminder-worker"
    RECURRENCE_WORKER = "recurrence-worker"
    SYSTEM = "system"


class TaskPriority(str, Enum):
    """Task priority levels."""
    HIGH = "HIGH"
    MEDIUM = "MEDIUM"
    LOW = "LOW"


class TaskStatus(str, Enum):
    """Task status values."""
    TODO = "TODO"
    IN_PROGRESS = "IN_PROGRESS"
    DONE = "DONE"


class RecurrenceFrequency(str, Enum):
    """Recurrence rule frequencies."""
    DAILY = "DAILY"
    WEEKLY = "WEEKLY"
    MONTHLY = "MONTHLY"
    YEARLY = "YEARLY"


class ReminderCancelReason(str, Enum):
    """Reasons why a reminder was cancelled."""
    TASK_DELETED = "task-deleted"
    DUE_DATE_REMOVED = "due-date-removed"
    TASK_COMPLETED = "task-completed"


# ============================================================================
# SHARED MODELS
# ============================================================================

class RecurrenceRule(BaseModel):
    """Defines how a task recurs."""
    frequency: RecurrenceFrequency
    interval: int = Field(default=1, gt=0, description="How often to recur (e.g., 1=every, 2=every other)")
    days_of_week: list[int] | None = Field(default=None, ge=0, le=6)
    end_date: datetime | None = None
    end_after_occurrences: int | None = Field(default=None, gt=0)
    cron_expression: str | None = Field(default=None, max_length=100)

    @field_validator("days_of_week")
    @classmethod
    def validate_days_of_week(cls, v, info):
        """Ensure days_of_week is only provided for WEEKLY frequency."""
        if v is not None and info.data.get("frequency") != RecurrenceFrequency.WEEKLY:
            raise ValueError("days_of_week is only valid for WEEKLY frequency")
        return v


class FieldChange(BaseModel):
    """Represents a single field change in TaskUpdatedEvent."""
    old: dict | list | str | int | float | bool | None
    new: dict | list | str | int | float | bool | None


# ============================================================================
# PAYLOAD MODELS
# ============================================================================

class TaskCreatedPayload(BaseModel):
    """Payload for task-created event."""
    task_id: UUID
    title: str = Field(..., min_length=1, max_length=500)
    description: str | None = None
    priority: TaskPriority
    status: TaskStatus
    due_date: datetime | None = None
    recurrence: RecurrenceRule | None = None
    tags: list[str] = Field(default_factory=list)
    user_id: UUID
    created_at: datetime


class TaskUpdatedPayload(BaseModel):
    """Payload for task-updated event."""
    task_id: UUID
    changed_fields: list[str]
    changes: dict[str, FieldChange]
    updated_at: datetime


class TaskDeletedPayload(BaseModel):
    """Payload for task-deleted event."""
    task_id: UUID
    user_id: UUID
    deleted_at: datetime


class TaskCompletedPayload(BaseModel):
    """Payload for task-completed event."""
    task_id: UUID
    user_id: UUID
    completed_at: datetime
    was_overdue: bool


class ReminderScheduledPayload(BaseModel):
    """Payload for reminder-scheduled event."""
    task_id: UUID
    reminder_id: UUID
    due_date: datetime
    reminder_time: datetime
    recurrence: RecurrenceRule | None = None


class ReminderTriggeredPayload(BaseModel):
    """Payload for reminder-triggered event."""
    task_id: UUID
    reminder_id: UUID
    task_title: str
    user_id: UUID
    due_date: datetime
    is_overdue: bool


class ReminderCancelledPayload(BaseModel):
    """Payload for reminder-cancelled event."""
    task_id: UUID
    reminder_id: UUID
    reason: ReminderCancelReason


class RecurrenceInstanceCreatedPayload(BaseModel):
    """Payload for recurrence-instance-created event."""
    parent_task_id: UUID
    instance_task_id: UUID
    scheduled_date: datetime
    recurrence_rule: RecurrenceRule


# ============================================================================
# BASE EVENT
# ============================================================================

class BaseEvent(BaseModel):
    """Base event that all events extend."""
    event_id: UUID = Field(default_factory=uuid4)
    event_type: EventType
    event_version: Literal["v1"] = "v1"
    timestamp: datetime = Field(default_factory=lambda: datetime.now(timezone.utc))
    correlation_id: UUID = Field(default_factory=uuid4)
    producer: EventProducer
    payload: (
        TaskCreatedPayload
        | TaskUpdatedPayload
        | TaskDeletedPayload
        | TaskCompletedPayload
        | ReminderScheduledPayload
        | ReminderTriggeredPayload
        | ReminderCancelledPayload
        | RecurrenceInstanceCreatedPayload
    )


# ============================================================================
# CONCRETE EVENTS
# ============================================================================

class TaskCreatedEvent(BaseEvent):
    """Event emitted when a new task is created."""
    event_type: Literal[EventType.TASK_CREATED] = EventType.TASK_CREATED
    payload: TaskCreatedPayload


class TaskUpdatedEvent(BaseEvent):
    """Event emitted when a task is updated."""
    event_type: Literal[EventType.TASK_UPDATED] = EventType.TASK_UPDATED
    payload: TaskUpdatedPayload


class TaskDeletedEvent(BaseEvent):
    """Event emitted when a task is deleted."""
    event_type: Literal[EventType.TASK_DELETED] = EventType.TASK_DELETED
    payload: TaskDeletedPayload


class TaskCompletedEvent(BaseEvent):
    """Event emitted when a task is marked complete."""
    event_type: Literal[EventType.TASK_COMPLETED] = EventType.TASK_COMPLETED
    payload: TaskCompletedPayload


class ReminderScheduledEvent(BaseEvent):
    """Event emitted when a reminder is scheduled."""
    event_type: Literal[EventType.REMINDER_SCHEDULED] = EventType.REMINDER_SCHEDULED
    payload: ReminderScheduledPayload


class ReminderTriggeredEvent(BaseEvent):
    """Event emitted when a reminder should be triggered."""
    event_type: Literal[EventType.REMINDER_TRIGGERED] = EventType.REMINDER_TRIGGERED
    payload: ReminderTriggeredPayload


class ReminderCancelledEvent(BaseEvent):
    """Event emitted when a reminder is cancelled."""
    event_type: Literal[EventType.REMINDER_CANCELLED] = EventType.REMINDER_CANCELLED
    payload: ReminderCancelledPayload


class RecurrenceInstanceCreatedEvent(BaseEvent):
    """Event emitted when a recurring task instance is created."""
    event_type: Literal[EventType.RECURRENCE_INSTANCE_CREATED] = EventType.RECURRENCE_INSTANCE_CREATED
    payload: RecurrenceInstanceCreatedPayload


# Union type for all events
AnyEvent = Union[
    TaskCreatedEvent,
    TaskUpdatedEvent,
    TaskDeletedEvent,
    TaskCompletedEvent,
    ReminderScheduledEvent,
    ReminderTriggeredEvent,
    ReminderCancelledEvent,
    RecurrenceInstanceCreatedEvent,
]


# ============================================================================
# KAFKA TOPIC CONFIGURATION
# ============================================================================

class KafkaTopics:
    """Kafka topic names."""
    TASK_EVENTS = "task-events"
    REMINDERS = "reminders"
    TASK_UPDATES = "task-updates"
    DEAD_LETTER = "dead-letter"


class TopicConfiguration:
    """Kafka topic configuration settings."""

    # Default settings for all topics
    DEFAULT_PARTITIONS = 3
    DEFAULT_REPLICATION_FACTOR = 2
    DEFAULT_RETENTION_MS = 7 * 24 * 60 * 60 * 1000  # 7 days

    # Topic-specific settings
    TOPIC_CONFIGS = {
        KafkaTopics.TASK_EVENTS: {
            "partitions": 3,
            "replication_factor": 2,
            "retention_ms": 7 * 24 * 60 * 60 * 1000,  # 7 days
            "cleanup_policy": "delete",
        },
        KafkaTopics.REMINDERS: {
            "partitions": 3,
            "replication_factor": 2,
            "retention_ms": 7 * 24 * 60 * 60 * 1000,  # 7 days
            "cleanup_policy": "delete",
        },
        KafkaTopics.TASK_UPDATES: {
            "partitions": 3,
            "replication_factor": 2,
            "retention_ms": 7 * 24 * 60 * 60 * 1000,  # 7 days
            "cleanup_policy": "delete",
        },
        KafkaTopics.DEAD_LETTER: {
            "partitions": 3,
            "replication_factor": 2,
            "retention_ms": 30 * 24 * 60 * 60 * 1000,  # 30 days
            "cleanup_policy": "delete",
        },
    }


# ============================================================================
# EVENT TO TOPIC MAPPING
# ============================================================================

EVENT_TO_TOPIC = {
    EventType.TASK_CREATED: KafkaTopics.TASK_EVENTS,
    EventType.TASK_UPDATED: KafkaTopics.TASK_UPDATES,
    EventType.TASK_DELETED: KafkaTopics.TASK_EVENTS,
    EventType.TASK_COMPLETED: KafkaTopics.TASK_UPDATES,
    EventType.REMINDER_SCHEDULED: KafkaTopics.REMINDERS,
    EventType.REMINDER_TRIGGERED: KafkaTopics.REMINDERS,
    EventType.REMINDER_CANCELLED: KafkaTopics.REMINDERS,
    EventType.RECURRENCE_INSTANCE_CREATED: KafkaTopics.TASK_EVENTS,
}


def get_topic_for_event(event_type: EventType) -> str:
    """Get the Kafka topic name for a given event type."""
    return EVENT_TO_TOPIC[event_type]


# ============================================================================
# EVENT HELPERS
# ============================================================================

def create_task_created_event(
    task_id: UUID,
    title: str,
    user_id: UUID,
    description: str | None = None,
    priority: TaskPriority = TaskPriority.MEDIUM,
    status: TaskStatus = TaskStatus.TODO,
    due_date: datetime | None = None,
    recurrence: RecurrenceRule | None = None,
    tags: list[str] | None = None,
) -> TaskCreatedEvent:
    """Helper to create a task-created event."""
    return TaskCreatedEvent(
        producer=EventProducer.CHAT_API,
        payload=TaskCreatedPayload(
            task_id=task_id,
            title=title,
            description=description,
            priority=priority,
            status=status,
            due_date=due_date,
            recurrence=recurrence,
            tags=tags or [],
            user_id=user_id,
            created_at=datetime.now(timezone.utc),
        ),
    )


def extract_task_id(event: BaseEvent) -> UUID | None:
    """Extract task_id from an event if present."""
    payload_dict = event.payload.model_dump()
    return payload_dict.get("task_id")


# ============================================================================
# MOCK EVENT PUBLISHER (For Testing Without Kafka)
# ============================================================================

class MockEventPublisher:
    """
    In-memory event publisher for development/testing without Kafka.

    This allows testing event contracts without deploying Kafka infrastructure.
    Events are stored in memory and can be retrieved for validation.
    """

    def __init__(self):
        self._events: dict[str, list[BaseEvent]] = {
            KafkaTopics.TASK_EVENTS: [],
            KafkaTopics.REMINDERS: [],
            KafkaTopics.TASK_UPDATES: [],
            KafkaTopics.DEAD_LETTER: [],
        }

    async def publish(self, event: BaseEvent) -> None:
        """Publish an event to the appropriate topic (in-memory)."""
        topic = get_topic_for_event(event.event_type)
        self._events[topic].append(event)

    async def get_events(
        self, topic: str | None = None, event_type: EventType | None = None
    ) -> list[BaseEvent]:
        """Retrieve events from the mock broker."""
        if topic:
            events = self._events.get(topic, [])
        else:
            events = [
                e for events in self._events.values() for e in events
            ]

        if event_type:
            events = [e for e in events if e.event_type == event_type]

        return events

    def clear(self) -> None:
        """Clear all stored events."""
        for topic in self._events:
            self._events[topic].clear()

    def get_event_count(self, topic: str | None = None) -> int:
        """Get the count of stored events."""
        if topic:
            return len(self._events.get(topic, []))
        return sum(len(events) for events in self._events.values())


# Singleton instance for testing
_mock_publisher = MockEventPublisher()


def get_mock_publisher() -> MockEventPublisher:
    """Get the singleton mock event publisher instance."""
    return _mock_publisher
```

---

## Usage Example

```python
from uuid import uuid4
from datetime import datetime, timezone

from backend.services.events.schemas import (
    TaskCreatedEvent,
    EventType,
    EventProducer,
    get_topic_for_event,
    get_mock_publisher,
)

# Create a new event
event = TaskCreatedEvent(
    producer=EventProducer.CHAT_API,
    payload={
        "task_id": uuid4(),
        "title": "Complete Phase 5 documentation",
        "description": "Create comprehensive architecture docs",
        "priority": "HIGH",
        "status": "TODO",
        "due_date": None,
        "recurrence": None,
        "tags": ["documentation", "phase-5"],
        "user_id": uuid4(),
        "created_at": datetime.now(timezone.utc),
    },
)

# Serialize to JSON for Kafka
event_json = event.model_dump(mode="json")

# Get target topic
topic = get_topic_for_event(EventType.TASK_CREATED)  # "task-events"

# Using mock publisher for testing (without Kafka)
mock = get_mock_publisher()
await mock.publish(event)
print(f"Published {event.event_type} to {topic}")
```

File: specs/009-kafka-events/contracts/event-schemas.ts
# Kafka Event Streaming - Type Definitions

> **Language**: TypeScript 5+
> **Purpose**: Event schema contracts for type safety and validation
> **Status**: Production Ready | **Deployment**: Not Required

```typescript
/**
 * Base Event Contract
 * All events MUST extend this interface
 */
export interface BaseEvent {
  /** Unique identifier for this event instance (UUID v4) */
  event_id: string;

  /** Type of event - determines payload structure */
  event_type: EventType;

  /** Schema version for compatibility */
  event_version: 'v1';

  /** Event creation timestamp in ISO-8601 UTC */
  timestamp: string;

  /** Correlation ID for distributed tracing (UUID v4) */
  correlation_id: string;

  /** Service that produced this event */
  producer: EventProducer;

  /** Event-specific payload data */
  payload: unknown;
}

/**
 * All possible event types in the system
 */
export type EventType =
  | 'task-created'
  | 'task-updated'
  | 'task-deleted'
  | 'task-completed'
  | 'reminder-scheduled'
  | 'reminder-triggered'
  | 'reminder-cancelled'
  | 'recurrence-instance-created';

/**
 * Services that can produce events
 */
export type EventProducer =
  | 'chat-api'
  | 'reminder-worker'
  | 'recurrence-worker'
  | 'system';

// ============================================================================
// TASK EVENTS
// ============================================================================

/**
 * Event emitted when a new task is created
 */
export interface TaskCreatedEvent extends BaseEvent {
  event_type: 'task-created';
  payload: TaskCreatedPayload;
}

export interface TaskCreatedPayload {
  task_id: string;
  title: string;
  description: string | null;
  priority: TaskPriority;
  status: TaskStatus;
  due_date: string | null;           // ISO-8601
  recurrence: RecurrenceRule | null;
  tags: string[];
  user_id: string;
  created_at: string;
}

/**
 * Event emitted when a task is updated
 * Only contains changed fields for efficiency
 */
export interface TaskUpdatedEvent extends BaseEvent {
  event_type: 'task-updated';
  payload: TaskUpdatedPayload;
}

export interface TaskUpdatedPayload {
  task_id: string;
  changed_fields: string[];
  changes: Record<string, FieldChange>;
  updated_at: string;
}

export interface FieldChange {
  old: unknown;
  new: unknown;
}

/**
 * Event emitted when a task is deleted
 */
export interface TaskDeletedEvent extends BaseEvent {
  event_type: 'task-deleted';
  payload: TaskDeletedPayload;
}

export interface TaskDeletedPayload {
  task_id: string;
  user_id: string;
  deleted_at: string;
}

/**
 * Event emitted when a task is marked as complete
 */
export interface TaskCompletedEvent extends BaseEvent {
  event_type: 'task-completed';
  payload: TaskCompletedPayload;
}

export interface TaskCompletedPayload {
  task_id: string;
  user_id: string;
  completed_at: string;
  was_overdue: boolean;
}

// ============================================================================
// REMINDER EVENTS
// ============================================================================

/**
 * Event emitted when a reminder is scheduled
 */
export interface ReminderScheduledEvent extends BaseEvent {
  event_type: 'reminder-scheduled';
  payload: ReminderScheduledPayload;
}

export interface ReminderScheduledPayload {
  task_id: string;
  reminder_id: string;
  due_date: string;                  // ISO-8601
  reminder_time: string;             // ISO-8601
  recurrence: RecurrenceRule | null;
}

/**
 * Event emitted when a reminder should be triggered
 * Produced by reminder worker at scheduled time
 */
export interface ReminderTriggeredEvent extends BaseEvent {
  event_type: 'reminder-triggered';
  payload: ReminderTriggeredPayload;
}

export interface ReminderTriggeredPayload {
  task_id: string;
  reminder_id: string;
  task_title: string;
  user_id: string;
  due_date: string;
  is_overdue: boolean;
}

/**
 * Event emitted when a reminder is cancelled
 * (e.g., task deleted, due_date removed)
 */
export interface ReminderCancelledEvent extends BaseEvent {
  event_type: 'reminder-cancelled';
  payload: ReminderCancelledPayload;
}

export interface ReminderCancelledPayload {
  task_id: string;
  reminder_id: string;
  reason: 'task-deleted' | 'due-date-removed' | 'task-completed';
}

// ============================================================================
// RECURRENCE EVENTS
// ============================================================================

/**
 * Event emitted when a new instance of a recurring task is created
 */
export interface RecurrenceInstanceCreatedEvent extends BaseEvent {
  event_type: 'recurrence-instance-created';
  payload: RecurrenceInstanceCreatedPayload;
}

export interface RecurrenceInstanceCreatedPayload {
  parent_task_id: string;
  instance_task_id: string;
  scheduled_date: string;           // ISO-8601
  recurrence_rule: RecurrenceRule;
}

// ============================================================================
// SHARED TYPES
// ============================================================================

export type TaskPriority = 'HIGH' | 'MEDIUM' | 'LOW';
export type TaskStatus = 'TODO' | 'IN_PROGRESS' | 'DONE';

export interface RecurrenceRule {
  frequency: 'DAILY' | 'WEEKLY' | 'MONTHLY' | 'YEARLY';
  interval: number;                 // e.g., 1 = every, 2 = every other
  days_of_week?: number[];          // 0-6 (Sunday-Saturday), for WEEKLY
  end_date?: string;                // ISO-8601, optional
  end_after_occurrences?: number;   // Optional
  cron_expression?: string;         // For advanced scheduling
}

// ============================================================================
// EVENT UTILITIES
// ============================================================================

/**
 * Type guard to check if an event is of a specific type
 */
export function isEventType<T extends BaseEvent>(
  event: BaseEvent,
  eventType: T['event_type']
): event is T {
  return event.event_type === eventType;
}

/**
 * Validate event structure
 * Throws if event is malformed
 */
export function validateEvent(event: unknown): BaseEvent {
  // TODO: Implement Zod schema validation
  // This would use Zod schemas to validate the event structure
  return event as BaseEvent;
}

/**
 * Extract task_id from any event if present
 */
export function extractTaskId(event: BaseEvent): string | null {
  const payload = event.payload as Record<string, unknown>;
  return (payload.task_id as string) ?? null;
}

// ============================================================================
// ZOD SCHEAS (For Runtime Validation)
// ============================================================================

import { z } from 'zod';

export const EventProducerSchema = z.enum([
  'chat-api',
  'reminder-worker',
  'recurrence-worker',
  'system'
]);

export const EventTypeSchema = z.enum([
  'task-created',
  'task-updated',
  'task-deleted',
  'task-completed',
  'reminder-scheduled',
  'reminder-triggered',
  'reminder-cancelled',
  'recurrence-instance-created'
]);

export const BaseEventSchema = z.object({
  event_id: z.string().uuid(),
  event_type: EventTypeSchema,
  event_version: z.literal('v1'),
  timestamp: z.string().datetime(),
  correlation_id: z.string().uuid(),
  producer: EventProducerSchema,
  payload: z.unknown()
});

export const TaskPrioritySchema = z.enum(['HIGH', 'MEDIUM', 'LOW']);
export const TaskStatusSchema = z.enum(['TODO', 'IN_PROGRESS', 'DONE']);

export const RecurrenceRuleSchema = z.object({
  frequency: z.enum(['DAILY', 'WEEKLY', 'MONTHLY', 'YEARLY']),
  interval: z.number().int().positive(),
  days_of_week: z.array(z.number().int().min(0).max(6)).optional(),
  end_date: z.string().datetime().optional(),
  end_after_occurrences: z.number().int().positive().optional(),
  cron_expression: z.string().optional()
});

export const TaskCreatedPayloadSchema = z.object({
  task_id: z.string().uuid(),
  title: z.string().min(1).max(500),
  description: z.string().nullable(),
  priority: TaskPrioritySchema,
  status: TaskStatusSchema,
  due_date: z.string().datetime().nullable(),
  recurrence: RecurrenceRuleSchema.nullable(),
  tags: z.array(z.string()),
  user_id: z.string().uuid(),
  created_at: z.string().datetime()
});

export const TaskCreatedEventSchema: z.ZodType<TaskCreatedEvent> = BaseEventSchema.extend({
  event_type: z.literal('task-created'),
  payload: TaskCreatedPayloadSchema
});

// ============================================================================
// KAFKA TOPIC CONFIGURATIONS
// ============================================================================

export const KAFKA_TOPICS = {
  TASK_EVENTS: 'task-events',
  REMINDERS: 'reminders',
  TASK_UPDATES: 'task-updates',
  DEAD_LETTER: 'dead-letter'
} as const;

export type KafkaTopic = typeof KAFKA_TOPICS[keyof typeof KAFKA_TOPICS];

/**
 * Maps event types to their target Kafka topics
 */
export const EVENT_TO_TOPIC: Record<EventType, KafkaTopic> = {
  'task-created': KAFKA_TOPICS.TASK_EVENTS,
  'task-updated': KAFKA_TOPICS.TASK_UPDATES,
  'task-deleted': KAFKA_TOPICS.TASK_EVENTS,
  'task-completed': KAFKA_TOPICS.TASK_UPDATES,
  'reminder-scheduled': KAFKA_TOPICS.REMINDERS,
  'reminder-triggered': KAFKA_TOPICS.REMINDERS,
  'reminder-cancelled': KAFKA_TOPICS.REMINDERS,
  'recurrence-instance-created': KAFKA_TOPICS.TASK_EVENTS
};
```

---

## Usage Example

```typescript
import { TaskCreatedEvent, TaskCreatedEventSchema, validateEvent } from './event-schemas';

// Create a new event
const event: TaskCreatedEvent = {
  event_id: crypto.randomUUID(),
  event_type: 'task-created',
  event_version: 'v1',
  timestamp: new Date().toISOString(),
  correlation_id: crypto.randomUUID(),
  producer: 'chat-api',
  payload: {
    task_id: crypto.randomUUID(),
    title: 'Complete Phase 5 documentation',
    description: 'Create comprehensive architecture docs',
    priority: 'HIGH',
    status: 'TODO',
    due_date: null,
    recurrence: null,
    tags: ['documentation', 'phase-5'],
    user_id: 'user-123',
    created_at: new Date().toISOString()
  }
};

// Validate using Zod
const validated = TaskCreatedEventSchema.parse(event);

// Get the target topic for this event
import { EVENT_TO_TOPIC } from './event-schemas';
const topic = EVENT_TO_TOPIC[event.event_type]; // 'task-events'
```
